dist/tools/ethos/ethos.c

```diff
@@ -267,6 +277,134 @@ static void _clear_neighbor_cache(const char *ifname)
     }
 }
 
+int parse_tcp_arg(char *name, char **host, char **port)
+{
+    char *host_str;
+    char *port_str;
+
+    // Remove 'tcp:'
+    name = &name[sizeof(TCP_DEV) -1];
+
+    // extract hostname
+    host_str = strtok(name, ":");
+
+    // extract port
+    port_str = strtok(NULL, ":");
+    if (!port_str) {
+        port_str = IOTLAB_TCP_PORT;
+    }
+
+    *host = host_str;
+    *port = port_str;
+
+    return 0;
+}
+
+/* Adapted from 'getaddrinfo' manpage example */
+int tcp_connect(char *host, char *port)
+{
+    int sfd = -1;
+    struct addrinfo hints, *result, *rp;
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+
+    int s = getaddrinfo(host, port, &hints, &result);
+    if (s) {
+        fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(s));
+        return -1;
+    }
+
+    /* getaddrinfo() returns a list of address structures.
+       Try each address until we successfully connect(2).
+       If socket(2) (or connect(2)) fails, we (close the socket
+       and) try the next address. */
+    for (rp = result; rp != NULL; rp = rp->ai_next) {
+        sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+        if (sfd == -1)
+            continue;
+
+        if (connect(sfd, rp->ai_addr, rp->ai_addrlen) != -1)
+            break;
+
+        close(sfd);
+    }
+
+    freeaddrinfo(result);
+
+    if (rp == NULL) {
+        fprintf(stderr, "Could not connect to '%s:%s'\n", host, port);
+        return -1;
+    }
+
+    return sfd;
+}
+
+int set_socket_timeout(int sfd)
+{
+    struct timeval timeout = {
+        .tv_sec = 0,
+        .tv_usec = TTY_TIMEOUT_MS * 1000,
+    };
+
+    if (setsockopt(sfd, SOL_SOCKET, SO_RCVTIMEO,
+                (char *)&timeout, sizeof(timeout)) == -1) {
+        perror("setsockopt failed\n");
+        return 1;
+    }
+
+    if (setsockopt(sfd, SOL_SOCKET, SO_SNDTIMEO,
+                (char *)&timeout, sizeof(timeout)) == -1) {
+        perror("setsockopt failed\n");
+        return 1;
+    }
+    return 0;
+}
+
+int open_tcp_connection(char *name)
+{
+    char *host;
+    char *port;
+
+    int ret = parse_tcp_arg(name, &host, &port);
+    if (ret) {
+        fprintf(stderr, "Error while parsing tcp arguments\n");
+        return -1;
+    }
+
+    int sfd = tcp_connect(host, port);
+    if (set_socket_timeout(sfd)) {
+        fprintf(stderr, "Error while setting socket options\n");
+        return -1;
+    }
+    return sfd;
+}
+
+int open_serial_link(char *name)
+{
+    int serial_fd = open(name, O_RDWR | O_NOCTTY | O_SYNC);
+
+    if (serial_fd < 0) {
+        fprintf(stderr, "Error opening serial device %s\n", name);
+        return -1;
+    }
+
+    set_serial_attribs(serial_fd, B115200, 0);
+    set_blocking(serial_fd, 1);
+
+    return serial_fd;
+}
+
+int open_serial(char *name)
```

Comment:
The function name could be improved:
`_open_connection` here and more specific names for serial and tcp connections (above): `_open_serial_connection` and `_open_tcp_connection`